apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/AbstractSamhsaMatcher.java

```diff
@@ -0,0 +1,378 @@
+    } catch (FHIRException e) {
+      /*
+       * This will only be thrown if the DiagnosisComponent doesn't have a
+       * CodeableConcept, which isn't how we build ours.
+       */
+      throw new BadCodeMonkeyException(e);
+    }
+  }
+
+  /**
+   * Checks that for the specified {@link CodeableConcept}, the Codings (if any) within, contain a
+   * blacklisted MHSA procedure code. If any of the systemms within th {@link CodeableConcept} are
+   * not known/expected, returns {@code true} and assuems the system is SAMHSA as a safety fallback.
+   *
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains any
+   *     {@link Coding}s that match any of the {@link #cptCodes} or has unknown coding systems,
+   *     <code>false</code> otherwise.
+   */
+  protected boolean containsSamhsaProcedureCode(CodeableConcept procedureConcept) {
+    return !procedureConcept.getCoding().isEmpty()
+        && (hasHcpcsSystemAndSamhsaCptCode(procedureConcept)
+            || !containsOnlyKnownSystems(procedureConcept));
+  }
+
+  /**
+   * Checks that for a {@link CodeableConcept} the {@link Coding}s contain a HCPCS system and at
+   * least one blacklisted CPT code.
+   *
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains a HCPCS
+   *     system as well as a CPT samhsa coding, (matched with {@link #cptCodes}), <code>false</code>
+   *     otherwise.
+   */
+  @VisibleForTesting
+  boolean hasHcpcsSystemAndSamhsaCptCode(CodeableConcept procedureConcept) {
+    /*
+     * Note: CPT codes represent a subset of possible HCPCS codes (but are the only
+     * subset that we blacklist from).
+     */
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream().map(Coding::getSystem).collect(Collectors.toSet());
+
+    // If no HCPCS system, it may be DME
+    return codingSystems.contains(TransformerConstants.CODING_SYSTEM_HCPCS)
+        && procedureConcept.getCoding().stream().anyMatch(this::isSamhsaCptCode);
+  }
+
+  /**
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains at least
+   *     one {@link Coding} and only contains {@link Coding}s that have known coding systems <code>
+   *     false</code> otherwise
+   */
+  @VisibleForTesting
+  boolean containsOnlyKnownSystems(CodeableConcept procedureConcept) {
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream().map(Coding::getSystem).collect(Collectors.toSet());
+
+    String hcpcsCdSystem = CCWUtils.calculateVariableReferenceUrl(CcwCodebookVariable.HCPCS_CD);
+
+    // Valid system url for productOrService coding
+    Set<String> hcpcsSystem = Set.of(TransformerConstants.CODING_SYSTEM_HCPCS);
```

Comment:
`hcpcsSystem` looks like a constant.  It could just be a `static final` ImmutableSet.